# NumPy 3 – Rechnen und Datentypen

**Szenario:** Bei der *Kiez Kaffee GmbH* wird heute gerechnet: Preise, Mengen, Mehrwertsteuer, Lagerbestände.
Dabei lernst du, warum NumPy so schnell rechnet und warum der **Datentyp** eines Arrays manchmal über richtig oder falsch entscheidet.

**Heute:** Potenzen, Wurzeln und Logarithmen in NumPy · Vektorisierung · ufuncs · Datentypen · typische Fallstricke

In [1]:
import numpy as np

# A – Brücke zur Mathe: Potenzen, Wurzeln, Logarithmen

## 1 – Die Mathe von heute Morgen in NumPy

| Mathe | NumPy | Beispiel |
|---|---|---|
| aⁿ | `a ** n` oder `np.power(a, n)` | `2 ** 3` → 8 |
| √a | `np.sqrt(a)` | `np.sqrt(16)` → 4 |
| ⁿ√a | `a ** (1/n)` | `27 ** (1/3)` → 3 |
| log₁₀(a) | `np.log10(a)` | `np.log10(1000)` → 3 |
| log₂(a) | `np.log2(a)` | `np.log2(8)` → 3 |
| ln(a) | `np.log(a)` | `np.log(np.e)` → 1 |
| eˣ | `np.exp(x)` | `np.exp(0)` → 1 |

**Achtung:** `np.log` ist der **natürliche** Logarithmus (Basis e), nicht log₁₀!

Das Besondere: Alle Funktionen arbeiten **mit ganzen Arrays**, nicht nur mit einzelnen Zahlen.

In [2]:
zahlen = np.array([1, 4, 9, 16, 25])

print("Quadrat  :", zahlen ** 2)
print("Wurzel   :", np.sqrt(zahlen))
print("3. Wurzel:", np.array([8, 27, 64]) ** (1/3))

Quadrat  : [  1  16  81 256 625]
Wurzel   : [1. 2. 3. 4. 5.]
3. Wurzel: [2. 3. 4.]


## 2 – Der Logarithmus als Umkehrung

Der Logarithmus beantwortet die Frage: **„Hoch wie viel?"**

- 2¹⁰ = 1024 → log₂(1024) = 10
- 10³ = 1000 → log₁₀(1000) = 3

Praktische Anwendung: log₁₀ verrät, **wie viele Stellen** eine Zahl hat.
Das ist nützlich, wenn Werte sehr unterschiedlich groß sind (3 Verkäufe vs. 12 000 Verkäufe).

In [ ]:
exponenten = np.arange(0, 11)
print(exponenten)
potenzen = 2 ** exponenten
print("2 hoch n :", potenzen)
print("log2     :", np.log2(potenzen))     # zurück zu den Exponenten

verkaeufe = np.array([3, 45, 820, 12000])
print("log10    :", np.log10(verkaeufe))
print("Stellen  :", np.floor(np.log10(verkaeufe)) + 1)      # nächstkleinere ganze Zahl

[ 0  1  2  3  4  5  6  7  8  9 10]
2 hoch n : [   1    2    4    8   16   32   64  128  256  512 1024]
log2     : [ 0.  1.  2.  3.  4.  5.  6.  7.  8.  9. 10.]
log10    : [0.47712125 1.65321251 2.91381385 4.07918125]
Stellen  : [0. 1. 2. 4.]


### Übung 1 – Potenzgesetze prüfen

👥 **Gemeinsame Übung**

1. Erstelle ein Array mit den Zehnerpotenzen **10⁰ bis 10⁶**.
2. Berechne davon log₁₀. Was fällt dir auf?
3. Prüfe das Potenzgesetz **aᵐ · aⁿ = aᵐ⁺ⁿ** für `a = 3`, `m = np.array([1, 2, 3])` und `n = np.array([2, 2, 2])`:
   Berechne beide Seiten und vergleiche sie mit `==`.
4. Prüfe **√(a · b) = √a · √b** für `a = np.array([4, 9, 16])` und `b = np.array([25, 4, 9])`.

In [47]:
# Dein Code hier
# 1
zehner = 10 ** np.arange(0,7)
print(zehner)
# 2
print(np.log10(zehner))
# 3
a = 3
m = np.array([1, 2, 3])
n = np.array([2, 2, 2])
print(a ** m * a ** n)
print(a ** (m + n))
print((a ** m * a ** n) == (a ** (m + n)))
print(3 ** 1 * 3 ** 2 == 3 ** (1 + 2))
print()
# 4
a = np.array([4, 9, 16])
b = np.array([25, 4, 9])
print(np.sqrt(a * b) == np.sqrt(a) * np.sqrt(b))

[      1      10     100    1000   10000  100000 1000000]
[0. 1. 2. 3. 4. 5. 6.]
[ 27  81 243]
[ 27  81 243]
[ True  True  True]
True

[ True  True  True]


# B – Vektorisierung: Rechnen ohne Schleife

## 3 – Array und Zahl

Alle Rechenzeichen `+`, `-`, `*`, `/`, `**`, `//`, `%` wirken **elementweise**:
Die Zahl wird mit **jedem** Element verrechnet.

In [5]:
preise = np.array([4.50, 3.20, 5.80, 2.90])   # Espresso, Crema, Filter, Kakao

print("+ 0,50 € :", preise + 0.50)
print("- 10 %   :", preise * 0.90)
print("in Cent  :", preise * 100)

+ 0,50 € : [5.  3.7 6.3 3.4]
- 10 %   : [4.05 2.88 5.22 2.61]
in Cent  : [450. 320. 580. 290.]


## 4 – Zwei Arrays gleicher Form

Haben zwei Arrays die **gleiche Form**, wird **Element für Element** gerechnet: das erste mit dem ersten, das zweite mit dem zweiten usw.

In [50]:
# Form = shape
mengen = np.array([120, 85, 60, 40])   # verkaufte Stück pro Produkt
print(preise)
print(np.shape(mengen))

umsatz = preise * mengen
print("Umsatz je Produkt:", umsatz)

[4.5 3.2 5.8 2.9]
(4,)
Umsatz je Produkt: [540. 272. 348. 116.]


Passen die Formen **nicht** zusammen, gibt es einen Fehler:

In [7]:
try:
    preise * np.array([1, 2, 3])
except ValueError as e:
    print("Fehler:", e)

Fehler: operands could not be broadcast together with shapes (4,) (3,) 


> 💡 **Kurz erwähnt:** Array und Zahl funktioniert, obwohl eine Zahl keine „Form" hat. Dahinter steckt **Broadcasting**, das lernst du in Notebook 4 genauer kennen.

## 5 – Warum ist das so schnell?

NumPy führt die Schleife **intern in C** aus, nicht in Python. Wir messen das mit `time.perf_counter()`:

In [51]:
import time

liste = list(range(1_000_000))
arr = np.arange(1_000_000)

start = time.perf_counter()
ergebnis_liste = [x * 2 for x in liste]
zeit_liste = time.perf_counter() - start

start = time.perf_counter()
ergebnis_arr = arr * 2
zeit_arr = time.perf_counter() - start

print(f"Liste: {zeit_liste * 1000:.1f} ms")
print(f"Array: {zeit_arr * 1000:.1f} ms")
print(f"NumPy ist ca. {zeit_liste / zeit_arr:.0f}-mal schneller")

Liste: 85.9 ms
Array: 4.4 ms
NumPy ist ca. 20-mal schneller


Die genauen Zahlen hängen von deinem Rechner ab, aber NumPy ist fast immer **deutlich** schneller.
Bei Millionen von Datensätzen macht das den Unterschied zwischen Sekunden und Minuten.

### Übung 2 – Tagesabrechnung

👥 **Gemeinsame Übung**

Nutze `preise` (netto, in €) und `mengen` von oben.

1. Berechne die **Bruttopreise** mit 19 % Mehrwertsteuer.
2. Berechne den **Bruttoumsatz** je Produkt (Bruttopreis × Menge).
3. Die Mengen vom Vortag waren `np.array([100, 90, 50, 45])`. Berechne die **Differenz** zum Vortag je Produkt.

In [57]:
# Dein Code hier
print(preise)   # netto, in EUR
print(mengen)
# 1
print(preise * 1.19)
# 2
print((preise * 1.19) * mengen)
# 3
vortag = np.array([100, 90, 50, 45])
print(mengen - vortag)

[4.5 3.2 5.8 2.9]
[120  85  60  40]
[5.355 3.808 6.902 3.451]
[642.6  323.68 414.12 138.04]
[20 -5 10 -5]


# C – Universelle Funktionen (ufuncs)

## 6 – Was ist eine ufunc?

Eine **ufunc** (*universal function*) ist eine NumPy-Funktion, die **elementweise** auf ein ganzes Array wirkt.
`np.sqrt`, `np.log10` oder `np.exp` aus Block A sind ufuncs.

| Gruppe | Beispiele |
|---|---|
| Mathematik | `np.sqrt`, `np.exp`, `np.log`, `np.log10`, `np.log2`, `np.power` |
| Runden | `np.round`, `np.floor`, `np.ceil`, `np.trunc` |
| Sonstiges | `np.abs`, `np.maximum`, `np.minimum` |

## 7 – Runden: vier Varianten

| Funktion | Bedeutung | 2.7 | -2.7 |
|---|---|---|---|
| `np.round(x, n)` | auf n Nachkommastellen runden auf nächste ganze Zahl | 3 | -3 |
| `np.floor(x)` | **ab**runden (nächstkleinere ganze Zahl) | 2 | -3 |
| `np.ceil(x)` | **auf**runden (nächstgrößere ganze Zahl) | 3 | -2 |
| `np.trunc(x)` | Nachkommastellen abschneiden | 2 | -2 |

Bei **negativen** Zahlen unterscheiden sich `floor` und `trunc`!

In [10]:
werte = np.array([2.7, 2.2, -2.7, -2.2])

print("round:", np.round(werte))
print("floor:", np.floor(werte))
print("ceil :", np.ceil(werte))
print("trunc:", np.trunc(werte))

print("Preise auf 2 Stellen:", np.round(preise * 1.19, 2))

round: [ 3.  2. -3. -2.]
floor: [ 2.  2. -3. -3.]
ceil : [ 3.  3. -2. -2.]
trunc: [ 2.  2. -2. -2.]
Preise auf 2 Stellen: [5.36 3.81 6.9  3.45]


> 💡 **Kurz erwähnt:** `np.round` rundet bei genau x,5 zur **geraden** Zahl („Banker's Rounding"): `np.round(2.5)` → 2, `np.round(3.5)` → 4.
> Das ist in der Statistik üblich.

## 8 – Betrag, Maximum und Minimum

- `np.abs(x)`: Betrag, macht negative Werte positiv (z. B. Abweichungen)
- `np.maximum(a, b)`: vergleicht **Element für Element** und nimmt jeweils den größeren Wert
- `np.minimum(a, b)`: nimmt jeweils den kleineren

`np.maximum` ist praktisch für **Untergrenzen**: `np.maximum(werte, 0)` ersetzt alles unter 0 durch 0.

In [11]:
soll = np.array([100, 80, 60, 50])
ist  = np.array([ 92, 85, 60, 41])

print("Abweichung     :", ist - soll)
print("Betrag         :", np.abs(ist - soll))
print("mind. 50 Stück :", np.maximum(ist, 50))
print("max. 90 Stück  :", np.minimum(ist, 90))

Abweichung     : [-8  5  0 -9]
Betrag         : [8 5 0 9]
mind. 50 Stück : [92 85 60 50]
max. 90 Stück  : [90 85 60 41]


> 💡 **Kurz erwähnt – Operatoren sind auch ufuncs**
> `a + b` ist dasselbe wie `np.add(a, b)`, `a > b` dasselbe wie `np.greater(a, b)`.
> Die Funktionsschreibweise hat Zusatzoptionen, z. B. `out=`, um das Ergebnis direkt in ein vorhandenes Array zu schreiben und Speicher zu sparen.
> Im Alltag nimmst du aber die Operatoren.

> 💡 **Nicht verwechseln:** `np.maximum(a, b)` vergleicht zwei Arrays elementweise. `a.max()` sucht den größten Wert **in einem** Array, das kommt in Notebook 4.

Bedarf an Kaffeebohnen je Filiale in kg:

In [12]:
bedarf_kg = np.array([12.3, 7.8, 15.0, 4.1])

### Übung 3 – Verpackung und Rundung

👥 **Gemeinsame Übung**

Die Filialen bestellen Kaffeebohnen in Kilogramm (`bedarf_kg`, Zelle oben).
Geliefert wird in **Säcken à 5 kg**.

1. Wie viele Säcke braucht jede Filiale? (Es darf nicht zu wenig sein!)
2. Wie viel kg bleiben pro Filiale übrig?
3. Jede Filiale bekommt **mindestens 2 Säcke**, auch wenn sie weniger braucht. Passe Aufgabe 1 an.

In [69]:
# Dein Code hier
print(bedarf_kg)
print()
# 1
print(np.ceil(bedarf_kg / 5))
saeke = np.ceil(bedarf_kg / 5)
# print(np.floor(bedarf_kg / 5 + 1))
print()
# 2
print(np.round(saeke * 5 - bedarf_kg, 1))
print()
# 3
print(np.maximum(saeke, 2))

[12.3  7.8 15.   4.1]

[3. 2. 3. 1.]

[2.7 2.2 0.  0.9]

[3. 2. 3. 2.]


# D – Datentypen (dtype)

## 9 – Was ist ein dtype?

Jedes Array hat **einen** Datentyp für **alle** Elemente. Er legt fest:

- welche Werte gespeichert werden können (ganze Zahlen, Kommazahlen, Wahrheitswerte, Text)
- welcher **Wertebereich** möglich ist
- wie viel **Speicher** jedes Element braucht

| Attribut | Bedeutung |
|---|---|
| `.dtype` | Datentyp |
| `.itemsize` | Bytes pro Element |
| `.nbytes` | Bytes für das ganze Array |

In [14]:
ganz = np.array([1, 2, 3])
komma = np.array([1.5, 2.0, 3.25])
wahr = np.array([True, False, True])
text = np.array(["Espresso", "Crema"])

for a in [ganz, komma, wahr, text]:
    print(a.dtype, "| Bytes pro Element:", a.itemsize)

int64 | Bytes pro Element: 8
float64 | Bytes pro Element: 8
bool | Bytes pro Element: 1
<U8 | Bytes pro Element: 32


## 10 – Wertebereiche: hier kommt 2ⁿ ins Spiel

Ganze Zahlen werden in **Bits** gespeichert. Mit n Bits gibt es **2ⁿ verschiedene Werte**.

| dtype | Bits | Anzahl Werte | Wertebereich |
|---|---|---|---|
| `uint8` | 8 | 2⁸ = 256 | 0 bis 255 |
| `int8` | 8 | 2⁸ = 256 | -128 bis 127 |
| `uint16` | 16 | 2¹⁶ = 65 536 | 0 bis 65 535 |
| `int32` | 32 | 2³² ≈ 4,3 Mrd. | ca. ±2,1 Mrd. |
| `int64` | 64 | 2⁶⁴ | ca. ±9,2 Trillionen |

- `u` = *unsigned* = **ohne Vorzeichen**, nur Werte ≥ 0
- ohne `u`: die 256 Werte werden auf negativ und positiv aufgeteilt

Für Kommazahlen gibt es `float32` (ca. 7 gültige Stellen) und `float64` (ca. 15 Stellen, Standard).

Mit `np.iinfo` kannst du die Grenzen nachschlagen:

In [15]:
for typ in [np.uint8, np.int8, np.uint16, np.int32]:
    info = np.iinfo(typ)
    print(f"{typ.__name__:7}: {info.min} bis {info.max}")

print("2 hoch 8 =", 2 ** 8, "| 2 hoch 16 =", 2 ** 16)

uint8  : 0 bis 255
int8   : -128 bis 127
uint16 : 0 bis 65535
int32  : -2147483648 bis 2147483647
2 hoch 8 = 256 | 2 hoch 16 = 65536


## 11 – dtype festlegen und umwandeln

**Beim Erstellen** mit `dtype=`:
```python
np.array([1, 2, 3], dtype=np.uint8)
np.zeros(5, dtype=np.int32)
```

**Nachträglich** mit `.astype()`. Wichtig: `astype` erzeugt immer eine **neue Kopie**, das Original bleibt unverändert.

In [16]:
bestand = np.array([12, 40, 7], dtype=np.uint8)
print(bestand, bestand.dtype)

bestand_gross = bestand.astype(np.int32)
print(bestand_gross, bestand_gross.dtype)
print("Original:", bestand.dtype)          # unverändert

[12 40  7] uint8
[12 40  7] int32
Original: uint8


> 💡 **Kurz erwähnt – Speicher:** Ein `int64` braucht 8 Byte, ein `uint8` nur 1 Byte. Bei Millionen von Werten spart der kleinere Typ viel Arbeitsspeicher.
> Für unsere Übungsdaten spielt das keine Rolle.

> 💡 **Kurz erwähnt – Text:** In älteren Tutorials steht `np.string_` oder `np.unicode_`. Diese Namen gibt es seit **NumPy 2.0 nicht mehr**, heute heißen sie `np.bytes_` und `np.str_`.

### Übung 4 – Den passenden Typ wählen

👥 **Gemeinsame Übung**

Wähle für jede Spalte den **kleinsten passenden** Integer-Typ (`uint8`, `int8`, `uint16`, `int16`, `uint32` …) und begründe kurz:

| Daten | Wertebereich |
|---|---|
| Alter der Kundinnen und Kunden | 16 bis 99 |
| Temperaturabweichung im Lager | -20 bis +20 |
| Lagerbestand in Packungen | 0 bis 50 000 |
| Kunden-IDs | 1 bis 3 Millionen |

Erzeuge dann für die ersten beiden je ein Beispiel-Array mit `dtype=` und prüfe mit `np.iinfo`, dass die Grenzen passen.

In [ ]:
# Dein Code hier
alter = np.array([16, 40, 99, dtype=np.uint8]) # max 99 -> uint8 (0...255)
temperatur = np.array([-20, 0, 20], dtype=np.int8) # negativ möglich, klein (Celsius) -> int(-128...127)
lager = np.array([0, 1000, 50000], dtype=uint16) # nie negativ, bis 50K -> uint16 (0... 65_535)
ku_id = np.array([1, 40, 3_000_000], dtype=np.uint32) # nie negativ, bis 3 Mio -> uint16 (0...~4,3 Mrd.)

# E – Typische Fallstricke

## 12 – Überlauf: wenn der Wertebereich nicht reicht

Wird beim Rechnen die Grenze eines Integer-Typs überschritten, **läuft der Wert über** und fängt von vorne an.
NumPy gibt dabei **keine Warnung** aus!

Stell dir einen Kilometerzähler vor, der nach 999 wieder bei 000 anfängt.

In [70]:
lager = np.array([250, 100, 30], dtype=np.uint8)
print("Bestand      :", lager)
print("+ 10 geliefert:", lager + 10)     # 250 + 10 = 260 passt nicht in uint8!

print("- 50 verkauft:", lager - 50)     # 30 - 50 = -20 geht bei uint8 nicht!

Bestand      : [250 100  30]
+ 10 geliefert: [  4 110  40]
- 50 verkauft: [200  50 236]


Aus 260 wird 4 (260 − 256) und aus −20 wird 236 (−20 + 256). Die Rechnung ist falsch, und niemand merkt es.

**Lösung:** Rechtzeitig einen größeren Typ wählen, z. B. `lager.astype(np.int32) + 10`.

## 13 – Kommazahl → ganze Zahl: es wird abgeschnitten

`astype(int)` **rundet nicht**, sondern schneidet die Nachkommastellen ab (wie `np.trunc`).
Wer kaufmännisch runden will, muss **vorher** `np.round` anwenden.

In [19]:
preise_cent = np.array([449.9, 319.6, 579.5])

print("astype    :", preise_cent.astype(int))
print("erst round:", np.round(preise_cent).astype(int))

astype    : [449 319 579]
erst round: [450 320 580]


## 14 – Automatische Typumwandlung (Upcasting)

Rechnet man mit verschiedenen Typen, wählt NumPy automatisch den **allgemeineren** Typ, damit nichts verloren geht:

| Kombination | Ergebnis |
|---|---|
| int + float | float |
| bool + int | int (True = 1, False = 0) |

Der zweite Fall ist sehr praktisch: Die **Summe einer Maske** zählt die `True`-Werte.

In [20]:
ganz = np.array([1, 2, 3])
print((ganz + 0.5).dtype)            # int + float -> float

print(np.array([1, 2.5, 3]).dtype)   # gemischte Liste -> float

maske = mengen > 70
print(maske, "-> Anzahl True:", maske.sum())

float64
float64
[ True  True False False] -> Anzahl True: 2


## 15 – Text in Zahlen-Arrays und der Typ `object`

Steht in einer Liste auch nur **ein** Text, wird das ganze Array zu Text. Dann kann man nicht mehr rechnen.
Bei ganz gemischten Inhalten (z. B. mit `None`) fällt NumPy auf `object` zurück: Das Array speichert dann nur Verweise auf Python-Objekte, **alle Vorteile von NumPy sind weg**.

In [21]:
gemischt = np.array([4.50, "3,20", 5.80])     # Komma statt Punkt -> Text!
print(gemischt, gemischt.dtype)

try:
    gemischt * 2
except Exception as e:
    print("Fehler:", type(e).__name__)

mit_luecke = np.array([4.50, None, 5.80])
print(mit_luecke, mit_luecke.dtype)

['4.5' '3,20' '5.8'] <U32
Fehler: TypeError
[4.5 None 5.8] object


Wenn dein Array plötzlich `<U…` (Text) oder `object` als dtype hat, stimmt meist etwas mit den **Rohdaten** nicht.
Für fehlende Werte gibt es in NumPy `np.nan`, das lernst du in Notebook 4.

## 16 – Kommazahlen sind nicht immer exakt

Computer speichern Kommazahlen im Binärsystem. Manche Dezimalzahlen wie 0,1 lassen sich dort nicht exakt darstellen,
genau wie 1/3 im Dezimalsystem nicht (0,333…). Deshalb:

In [22]:
print(0.1 + 0.2)
print(0.1 + 0.2 == 0.3)
print(np.isclose(0.1 + 0.2, 0.3))    # prüft "fast gleich"

0.30000000000000004
False
True


**Regel:** Kommazahlen nie mit `==` vergleichen, sondern mit `np.isclose`.

Code des Kollegen:

In [71]:
bestand = np.array([200, 180, 240], dtype=np.uint8)
lieferung = np.array([60, 20, 30], dtype=np.uint8)
neu = bestand + lieferung

### Übung 5 – Fehler im Lager finden

👥 **Gemeinsame Übung**

Ein Kollege hat den Code in der Zelle oben geschrieben. Das Ergebnis ist falsch.

1. Führe den Code aus. Welche Werte sind falsch und warum?
2. Korrigiere den Code so, dass das Ergebnis stimmt.
3. Der Durchschnittspreis pro Packung ist `np.array([4.35, 2.30, 1.15])`. Er soll in **ganze Cent** umgerechnet werden.
   Warum ist `(preis * 100).astype(int)` gefährlich? Schreibe es richtig.

In [75]:
# Dein Code hier
# 1
print(neu)
# 2
neu = bestand.astype(np.int32) + lieferung
print(neu)
# 3
preis = np.array([4.35, 2.30, 1.15])
print(preis * 100)
print(f"{preis[0] * 100:.15f}")


[260 200 270]
[260 200 270]
[435. 230. 115.]
434.999999999999943


# F – Übungen zum selbstständigen Arbeiten

Hier übst du allein oder zu zweit. Die Übungen sind nach **Themen** sortiert.
Bei jeder Übung steht, zu welchem Block sie gehört. Wenn du bei einer Aufgabe hängst, schau im passenden Block nach.

## 17 – Daten für die Übungen

Führe diese Zelle zuerst aus:

- `produkte`, `preise_netto`: Sortiment und Nettopreise in €
- `verkauf_mo`, `verkauf_di`: verkaufte Stück pro Produkt am Montag und Dienstag
- `jahresverkauf`: verkaufte Stück pro Produkt im ganzen Jahr
- `lager`: aktueller Lagerbestand pro Produkt (Achtung, Datentyp!)

In [25]:
produkte     = np.array(["Espresso", "Crema", "Filter", "Decaf", "Kakao", "Chai"])
preise_netto = np.array([4.20, 3.10, 5.40, 4.60, 2.80, 3.90])
verkauf_mo   = np.array([140, 95, 60, 12, 48, 30])
verkauf_di   = np.array([128, 110, 52, 18, 40, 35])
jahresverkauf = np.array([48200, 31500, 19800, 3100, 950, 12])
lager        = np.array([230, 180, 95, 40, 250, 12], dtype=np.uint8)

## 18 – Thema: Potenzen, Wurzeln, Logarithmen

### Übung 6 – Größenordnungen

📌 **Thema:** Potenzen, Wurzeln, Logarithmen (Block A)

Arbeite mit `jahresverkauf`.

1. Berechne log₁₀ der Jahresverkäufe.
2. Wie viele **Stellen** hat jede Zahl? (Tipp: Block A, Abschnitt 2)
3. Welche Produkte verkaufen sich im **fünfstelligen** Bereich? (Tipp: Maske)

In [26]:
# Dein Code hier

### Übung 7 – Verdopplung

📌 **Thema:** Potenzen, Wurzeln, Logarithmen (Block A)

Kiez Kaffee möchte wissen, wann sich die Verkäufe **verdoppeln**, wenn sie jedes Jahr um einen festen Prozentsatz wachsen.

Aus der Mathe: Bei Wachstumsfaktor q gilt qⁿ = 2, also **n = log(2) / log(q)**.

1. Berechne die Verdopplungszeit (in Jahren) für die Wachstumsraten `np.array([2, 5, 10, 20])` Prozent.
   (Tipp: Wachstumsfaktor q = 1 + Prozent / 100)
2. Runde auf **ganze Jahre auf**.
3. Prüfe für 10 %: Ist 1,1 hoch deinem Ergebnis aus Aufgabe 2 wirklich mindestens 2?

In [27]:
# Dein Code hier

## 19 – Thema: Vektorisierung

### Übung 8 – Zwei Tage im Vergleich

📌 **Thema:** Vektorisierung (Block B)

1. Berechne den **Nettoumsatz** je Produkt für Montag und Dienstag.
2. Berechne den **Bruttoumsatz** (19 % MwSt.) je Produkt für **beide Tage zusammen**.
3. Um wie viel **Prozent** hat sich die verkaufte Menge von Montag auf Dienstag verändert? (Runde auf 1 Stelle.)

In [28]:
# Dein Code hier

## 20 – Thema: ufuncs

### Übung 9 – Preisrunde

📌 **Thema:** ufuncs (Block C)

Kiez Kaffee möchte die **Bruttopreise** (19 % MwSt.) kundenfreundlich machen.

1. Berechne die Bruttopreise, gerundet auf 2 Stellen.
2. Runde die Bruttopreise auf **volle 10 Cent auf** (z. B. 3,69 → 3,70). Tipp: mal 10, `np.ceil`, durch 10.
3. Kein Produkt soll unter **3,50 €** kosten. Setze die Untergrenze mit einer ufunc aus Block C.

In [29]:
# Dein Code hier

### Übung 10 – Schwankungen

📌 **Thema:** ufuncs (Block C)

1. Berechne die **absolute** Abweichung der verkauften Menge zwischen Montag und Dienstag je Produkt.
2. Welche Produkte schwanken um **mehr als 10 Stück**?
3. Erstelle ein Array, das je Produkt die **höhere** der beiden Tagesmengen enthält.

In [30]:
# Dein Code hier

## 21 – Thema: Datentypen

### Übung 11 – Typen und Speicher

📌 **Thema:** Datentypen (Block D)

1. Gib für `preise_netto`, `verkauf_mo`, `lager` und `produkte` jeweils `dtype` und `itemsize` aus.
2. `verkauf_mo` braucht nie mehr als 1000. Wandle es in den **kleinsten passenden** Typ um.
3. Vergleiche `.nbytes` vorher und nachher. Wie viel Prozent Speicher sparst du?

In [31]:
# Dein Code hier

## 22 – Thema: Fallstricke

Lieferung von heute:

In [32]:
lieferung_heute = np.array([40, 30, 20, 20, 10, 50], dtype=np.uint8)

### Übung 12 – Lager nach Lieferung

📌 **Thema:** Fallstricke (Block E)

Heute kommt die Lieferung `lieferung_heute` (Zelle oben). Das Warenwirtschaftssystem liefert sie als `uint8`.

1. Berechne `lager + lieferung_heute` direkt. Welche Werte sind falsch? Woran erkennst du das?
2. Berechne den neuen Bestand **korrekt**.
3. Zähle mit einer Maske und `.sum()`, wie viele Produkte nach der Lieferung **mehr als 200** Packungen haben.

In [33]:
# Dein Code hier

Export aus der Kasse:

In [34]:
kasse = ["4,20", "3.10", "5,40"]

### Übung 13 – Daten aus der Kasse

📌 **Thema:** Fallstricke (Block E)

Die Kasse liefert Preise als Text, teilweise mit Komma (`kasse`, Zelle oben).

1. Erstelle daraus mit `np.array(kasse)` ein Array. Welchen dtype hat es?
2. Ersetze die Kommas durch Punkte (Tipp: List Comprehension mit `.replace(",", ".")`) und wandle das Ergebnis mit `.astype(float)` in Zahlen um.
3. Prüfe mit `np.isclose`, ob die Preise mit `preise_netto[:3]` übereinstimmen.

In [35]:
# Dein Code hier

---
**Morgen:** Prozent, Dreisatz und Wachstum mit NumPy: Summen und Durchschnitte, fehlende Werte, und wie NumPy Arrays unterschiedlicher Form zusammenrechnet (Broadcasting).